README.md

<img src="inst/application/www/SmaRPSticker.png" align="right" width="15%" height="15%"/>

# SmaRP: Smart Retirement Planning

Smart Retirement Planning (**SmaRP**) is a [Mirai Solutions](https://mirai-solutions.ch/) initiative designed to guide people working in Switzerland towards a strategic decision-making process for their retirement.

**SmaRP** is based on the [Swiss pension system](https://en.wikipedia.org/wiki/Pension_system_in_Switzerland) and reflects the complexity of its legal framework.
It is implemented as an [R Shiny](https://shiny.rstudio.com/) pension calculator web app, in the form of an R package.
The app features a flexible yet intuitive user interface with detailed personalization parameters and options.
This allows to interactively compute and display the evolution of the retirement funds over time, split into their contributing components.
A report including details regarding calculation methodology and approximations can be generated and downloaded.

Thanks to the open-source nature of the project, the underlying assumptions and calculations are fully disclosed.
Unlike other pension calculators, this makes results transparent, comparable, and reproducible.


## Using SmaRP

The **SmaRP** Shiny app is [deployed](gke#readme) to Google Cloud Platform
(using [Docker containers](https://www.docker.com/resources/what-container)) and
can be accessed at https://mirai-solutions.ch/gallery/smarp.

The R package **SmaRP** can be installed from GitHub with
<!-- argument build_vignettes not available anymore (r-lib/remotes#353), build_opts = "" for a full installation including vignettes  -->
``` r
remotes::install_github("miraisolutions/SmaRP", build_opts = "")
```
and used to serve the app locally from R via
``` r
SmaRP::launch_application()
```
**SmaRP** is developed using a [GitFlow](git-flow#readme) approach, hence the `master` branch always reflects the _latest_ [release](https://github.com/miraisolutions/SmaRP/releases) of the live app, whereas branch `develop` collects the latest delivered developments for the _next_ releases, which can be installed locally via
``` r
remotes::install_github("miraisolutions/SmaRP", "develop", build_opts = "")
```

Note that **SmaRP** is deployed using [version-stable](https://github.com/rocker-org/rocker-versioned#readme) images from the [Rocker project](https://www.rocker-project.org/). The target environment of the live app is currently bound to R 3.5.3. Therefore, the app is developed and tested with the corresponding version of R and packages, as opposed to the latest available versions.


## Details and key features

The evolution of the total retirement fund over time is computed by projecting the value of the occupational fund (2nd Pillar), the private fund (3rd Pillar) and the tax relief, thus deriving their contributions at the desired retirement age.

- _Contributions to the second Pillar_ are calculated from the salary and any additional voluntary purchases.

- _Contributions to the third Pillar_ are fully voluntary and repeated every year until retirement.

- _Tax savings_ are built as an additional fund where tax relieves from a certain year are used as contributions for the next. Tax relieves are calculated using an approximation of the given gross salary and other factors including: residence, civil status, number on kids, etc. 

**Results** of the calculation are available in **SmaRP** in 3 different ways:

1. a **graph** showing the projected funds over time until retirement
2. a **table** with details about projected quantities (also available for download)
3. a downloadable **PDF report** including detailed explanations of the calculation methodology and inputs.


### Assumptions and limitations

- Projected funds are computed using [continuously compounded returns](https://en.wikipedia.org/wiki/Compound_interest#Continuous_compounding) on annual basis.
- Constant interest rates are assumed throughout the working life.
- Inflation is not taken into account, although it can be proxied using the salary growth rate input.
- The retirement plan is valid for employees only, i.e. persons whose main income is a salary. Self-employed people do not belong to this category.
- The publicly managed pay-as-you-go system (1st Pillar) is not considered.
- All generated tax benefits are 100% reinvested as an additional fund, interpreting the same return as the private fund.
- In case of married couples with double income, the combined amount of all variables should be entered and a 50% income distribution is assumed.


### Source code

Core calculations behind this Shiny app have been implemented via several functions, collected in the main source files [SmaRP/R/core.R](R/core.R) and [SmaRP/R/TaxBenefits.R](R/TaxBenefit.R).

Documentation for the relevant exported functions used in the app is also provided and can be browsed via
``` r
help(package = "SmaRP")
```
On a functional level, the code is covered by extensive [unit tests](tests/testthat).


The source code for the app itself is available under  [SmaRP/inst/application](inst/application).


### Data sources and references

**SmaRP** reflects the Swiss pension system and uses corresponding legal parameters and data.
An overview of these components can be found at e.g. https://en.wikipedia.org/wiki/Pension_system_in_Switzerland.
A more detailed explanation is available on the Swiss [Federal Social Insurance Office](https://www.bsv.admin.ch/bsv/de/home/sozialversicherungen/ueberblick.html) website (in German).


Legal parameters in **SmaRP** are defined in [SmaRP/inst/application/global.R](inst/application/global.R), whereas data is stored under [SmaRP/inst/application/data](inst/application/data).


### Accuracy

While **SmaRP** was developed under the utmost care and diligence, Mirai Solutions does not guarantee for its accuracy and correctness. In addition, **SmaRP** is based on assumptions and projections (explained in a section above and in the PDF report) and as such computed figures should be understood as general references and do not hold any legal value.

Besides standard unit tests at the functional level, results from the **SmaRP** web app have been checked against other online free sources for second, third Pillars and tax calculators.

To keep testing and improving **SmaRP**, we encourage users to get back to us! Your feedback is always highly appreciated. You can use the issue tracker on GitHub to suggest enhancements or report problems, and reach out via email at info@mirai-solutions.com for any questions and comments.


NEWS.md
# SmaRP 1.3.0-9000

# SmaRP 1.3.0

### Changes

* The app now uses custom collapsible panels, removing the dependency on shinydashboardPlus (#113).
* The Dockerfile for the deployed app was improved, and is now based on R 3.5.3 (#91, #152).
* The total retirement fund text in the app was improved (#114).
* The PDF report now uses A4 as pages size (#131).
* The increment of interest rates in the app is now aligned to 0.1% (#142).
* The iframe-resizer version was upgraded to 4.2.1 (#144). 

### Maintenance

* Instructions and manifests for the GKE deployment are now included in the `gke` directory (#141).
* GitFlow and release details are now included in the `git-flow` directory (#143).
* The README was updated based on the new website integration of the deployed app (#143).

# SmaRP 1.2.0

### Changes

* The naming of pension funds was reviewed and aligned, and now reflects the official terminology (#103).
* The 3rd Pillar annual contribution in the app is now initialized with a non-zero value (#129).
* Negative input values are now re-set to zero, and maximum allowed values were consolidated (#129).
* `launch_application()` exposes the `launch.browser` argument of `shiny::runApp()` (#108).
* Minor updates to package vignettes (#101).

### Fixes

* Missing page title in the app (#118).
* Missing automatic cleanup of data created via configure (#119).
* App header background image re-centering upon page resize (#123).

### Maintenance

* Included `URL` and `BugReports` in DESCRIPTION file (#125).
* Updated `install_github()` README instructions due to un-supported `build_vignettes` argument (#120).

# SmaRP 1.1.1

### Fixes

* Bar-plot missing in the PDF report (#96).
* PDF report refers to wrong pillars (#105).
* Checkboxes and radio buttons don't render properly in Safari browser (#102).

### Maintenance

* Aligned plot labels with table headers in the report.
* Removed obsolete dependency on package webshot and PhantomJS (#97).

# SmaRP 1.1.0

* Retirement age automatically resets to 70 if a higher number is inserted (#27)
* Retirement age changes depending on the selected genre (#72)
* Updated legal parameters (01.01.2019) and tests (#55, #68)
* Removed user_manual.Rmd vignette created with (#60) and added new one (#65)
* SmaRP version shown in the app and the report (#57)
* Report review and updates (#82, #53, #89)
* Added sentence about not storing information in disclaimer, created "Download Data" button when table is displayed (#67)
* Made "c" in "calendar" capital, added tooltip to "Generate Report" button, made "Disclaimer" bold, rephrased first sentence and removed second (#77)
* "genre" changed to "gender" (#79)
* Fixed numeric input so that it doesn't crash when no value is inserted (#81, #28)
* Support responsive embedding via iframe-resizer (#69)
* Dynamic plot size and improved labels (no underscore, space between words) in the Shiny app (#85)

# SmaRP 1.0.1

## Patch release

* Report spell check (#49)
* Documentation reviewed (#59)
* Fixed tooltip automatic disappearing (#54)

# SmaRP 1.0.0

## First versioned release of SmaRP

Smart Retirement Planning **SmaRP** is a [Mirai Solutions](https://mirai-solutions.ch/) tool allowing people working in Switzerland to explore parameters and monetary measures related to retirement and pension.

**SmaRP** is based on the [Swiss pension system](https://en.wikipedia.org/wiki/Pension_system_in_Switzerland) and reflects the complexity of its legal framework.
It is implemented as an [R Shiny](https://shiny.rstudio.com/) pension calculator web app, in the form of an R package.
The app features a flexible yet intuitive user interface with detailed personalization parameters and options.
This allows to interactively compute and display the evolution of the retirement funds over time, split into their contributing components.
A report including details regarding calculation methodology and approximations can be generated and downloaded.

Thanks to the open-source nature of the project, the underlying assumptions and calculations are fully disclosed.
Unlike other pension calculators, this makes results transparent, comparable, and reproducible.


ui.R
library(shiny)
library(shinyBS)
library(shinythemes)
#library(shinydashboard)
source("external_inputs.R")
source("helper_texts.R")


# #FF9966 Mirai Orange
# #008cc3 Mirai blue
# #189bce background header
# #ea8b5b title orange

# UI
shinyUI( 
  fluidPage(
    # add "title" tag within the head of the html page
    titlePanel(title = NULL, windowTitle = "SmaRP: Smart Retirement Planning"),
    #themeSelector(),
    theme = "style.css",
    fluidRow(
      # tags$head(
      #   tags$style(HTML("hr {border-top: 1px solid #000000;}"))
      # ),
      fluidRow(img(src='mirai.png', align = "right",  height = "10%", width = "10%"), style = "margin-top: 10px;margin-right: 30px;margin-bottom: 10px;"),
      #titlePanel(
        fluidRow(
          id="head1",
          column(1, 
                 fluidRow(),
                 fluidRow(img(src='SmaRPStiker.png',  height = "110%", width = "110%"), style="margin:10px;")
                 ),
          column(6,
                 fluidRow(h2("SmaRP:")),
                 fluidRow(h3("Smart Retirement Planning"))
          ), # end of first column
          # column(4, 
          #        fluidRow(img(src='mirai.png', align = "right",  height = "70%", width = "70%"), style = "margin-top: 20px;")#, height = 100, width = 300
          #        #fluidRow(a(href="http://www.mirai-solutions.com", "mirai-solutions.com"),align="right",style = "margin-right: 10px;" )
          #        ), # end of second column
          # column(1) # empty column to have the logo no too much to the right
          style = "margin: 10px;" 
        ) #end of fluid row
      #)#, # end of title panel
      #hr()
    ), #end of FluidRow
    fluidRow(
      column(4,
             tabsetPanel(
               id = "inputs",
               type = "tabs",
               tabPanel(
                 title = "Generic Inputs", 
                 value = "Personal",
                 br(),
                 fluidRow( dateInput("Birthdate", label = h5("Birthdate"), value = "1980-12-31", format = "yyyy-mm-dd"),
                           style = "margin-left: 30px;"),
                 wellPanel(
                   checkboxInput("provideRetirementAge", "Desired Retirement Age (optional)", FALSE),
                   uiOutput("conditionalRetirementAge"),
                   style = "margin-left: 20px;margin-right: 20px;"
                 ),
                 hr(),
                 fluidRow( tags$h4("Private Pension Fund"), style = "margin-left: 30px;"),
                 #infoBoxOutput("ibox",width = 0.5),
                 fluidRow(
                   numericInput("CurrentP3", label = h5("Current assets"), value = 50000, step = 1000, min = 0),
                   bsTooltip("CurrentP3", IB$CurrentP3, placement = "right", options = list(container = "body")),
                   style = "margin-left: 30px;"),
                 fluidRow(
                   column( 6, numericInput("P3purchase", label = h5("Annual contribution"), value = 0, step = 500, min = 0),
                           bsTooltip("P3purchase", IB$P3purchase, placement = "right", options = list(container = "body"))),
                   column( 6, numericInput("returnP3", label = h5("Expected Return"), value = BVGMindestzinssatz, step = 0.001, min = 0, max = 0.25),
                           bsTooltip("returnP3", IB$returnP3, placement = "right", options = list(container = "body"))),
                   style = "margin-left: 20px;")
               ), #end of Personal Input tabPanel
               tabPanel(
                 title = "Particular cases", 
                 value = "Residence",
                 br(),
                 tabsetPanel(
                   id = "case",
                   type = "pills",
                   tabPanel(title = "General case", 
                            value = "General",
                            fluidRow( tags$h4("Tax Benefits"),style = "margin-left: 20px;"),
                            fluidRow( numericInput("TaxRelief", label = h5("Maximum Tax Relief"), value = 10000, step = 100, min = 0),
                                      bsTooltip("TaxRelief", IB$TaxRelief, placement = "right", options = list(container = "body")),
                                      style = "margin-left: 30px;"),
                            fluidRow( numericInput("TaxRate", label = h5("Marginal Tax Rate"), value = 0.1, step = 0.01, min = 0, max = 0.9),
                                      bsTooltip("TaxRate", IB$TaxRate, placement = "right", options = list(container = "body")),
                                      style = "margin-left: 30px;")
                            #,fluidRow( selectInput("currency", label = h5("Currency"), selected = "CHF", choices = currencies.list)
                            #          , style = "margin-left: 30px;")
                   ), # end General tabPanel
                   tabPanel(title = "Swiss case",
                            value = "Swiss",
                            conditionalPanel(condition= 'input.provideTaxRateSwiss==""',
                                             #                    fluidRow(tags$h4("Parameters for Tax Rate ecaluation"),style = "margin-left: 20px;"),
                                             fluidRow(column( 6, selectInput("postalcode", label = h5("Postal Code"),
                                                                             choices = PLZ.list,
                                                                             selected = "8001")),
                                                      column( 6, numericInput("NKids", label = h5("Number of Children"), value = 0, min = 0, max = 9),
                                                              bsTooltip("NKids", IB$NKids, placement = "right", options = list(container = "body"))),
                                                      style = "margin-left: 20px;"),
                                             conditionalPanel(condition= 'input.provideRetirementAge==""',
                                                              fluidRow(radioButtons("genre", label = NULL, inline = TRUE,
                                                                                    choices = list("Male" = "M", "Female" = "F"), 
                                                                                    selected = "M"),
                                                                       style = "margin-left: 30px;")
                                                              ), # end conditional panel
                                             fluidRow(radioButtons("rate_group", label = NULL, inline = TRUE, 
                                                                   choices = Rate_group.list, 
                                                                   selected = "A"),
                                                      bsTooltip("rate_group", IB$rate_group, placement = "right", options = list(container = "body")),
                                                      style = "margin-left: 30px;"), 
                                             #br(),
                                             fluidRow(radioButtons("churchtax", label = h5("Church affiliation"), inline = TRUE,
                                                                   choices = list("Y" = "Y", "N" = "N"), 
                                                                   selected = "N"),
                                                      style = "margin-left: 30px;")
                            ), # end conditional panel
                            hr(),
                            wellPanel(
                              checkboxInput("provideTaxRateSwiss", "Direct Tax Rate (optional)", FALSE),
                              uiOutput("conditionalInputSwiss"),
                              style = "margin-left: 20px;margin-right: 20px;"
                            ),
                            hr(),
                            fluidRow(tags$h4("Occupational Pension Fund (BVG)"),style = "margin-left: 20px;"),
                            fluidRow(
                              column( 6,numericInput("Salary", label = h5("Current Annual Salary"), value = 100000, step = 1000, min = 0),
                                      bsTooltip("Salary", IB$Salary, placement = "right", options = list(container = "body"))),
                              column( 6,numericInput("SalaryGrowthRate", label = h5("Expected salary growth rate"), value = 0.02, step = 0.001, min = 0, max = 0.1),
                                      bsTooltip("SalaryGrowthRate", IB$SalaryGrowthRate, placement = "right", options = list(container = "body"))),
                              style = "margin-left: 20px;"),
                            fluidRow(
                              column( 6, numericInput("CurrentP2", label = h5("Current BVG assets"), value = 100000, step = 1000, min = 0),
                                      bsTooltip("CurrentP2", IB$CurrentP2, placement = "right", options = list(container = "body"))),
                              column( 6, numericInput("P2interestRate", label = h5("Interest Rate % (optional)"), value = 100*BVGparams$BVGMindestzinssatz, step = 1, min = 100*BVGparams$BVGMindestzinssatz, max = 100),
                                      bsTooltip("P2interestRate", IB$P2interestRate, placement = "right", options = list(container = "body"))),
                              style = "margin-left: 20px;"),
                            #br(),
                            fluidRow(
                              column( 6, numericInput("P2purchase", label = h5("Voluntary purchases"), value = 0, step = 500, min = 0),
                                      bsTooltip("P2purchase", IB$P2purchase, placement = "right", options = list(container = "body"))),
                              column( 6, radioButtons("TypePurchase", label = NULL, inline = FALSE,  choices = Purchase.list),
                                      bsTooltip("TypePurchase", IB$TypePurchase, placement = "right", options = list(container = "body")), style = "margin-top: 20px;"),
                              style = "margin-left: 20px;"),
                            hr(),
                            fluidRow( 
                              wellPanel(
                                checkboxInput("login", "login as an admin", FALSE),
                                uiOutput("conditionalrefreshButton"),
                                uiOutput("conditionalrefreshText"),
                                style = "margin-left: 20px;margin-right: 20px;"
                              )
                              #,style = "margin-left: 30px;"
                              ),
                            br()
                   ) #  end Swiss tabPanel
                 ) # end of tabsetPanel
               ) # end of tab panel
             ) # end of tabsetPanel
      ), #end first column
      column(8, 
             tabsetPanel(
               type = "pills",
               tabPanel(title = "Plot", 
                        value = "Plot", 
                        #                   hr(),
                        fluidRow(align="center", verbatimTextOutput("Totals")),
                        fluidRow(htmlOutput("plot1")),
                        fluidRow(htmlOutput("plot2"), style = "margin-left: 60px;")#,
                        #fluidRow(align="center", textOutput('safeError'))
               ), # end tab Plot
               
               tabPanel(title = "Table", 
                        value = "Table", 
                        htmlOutput("table")
               ) # end tab Table
               
             ), # end tabsetPanel
             
             fluidRow(align="left", 
                      #Disclaimer
                      verbatimTextOutput("disclaimer")),
             # tags$head(tags$style("#Totals{
             #                      font-family:Helvetica;
             #                      color: blue;
             #                      font-size: 20px;
             #                      font-style: bold;
             #                      text-align: center;
             #                      }"),
             #           tags$style("#disclaimer{
             #                      font-family:Helvetica;
             #                      color: grey;
             #                      font-size: 12px;
             #                      text-align: left;
             #                      }")),
             fluidRow(align="left", 
                      #Add button to download report
                      downloadButton("report", "Generate report"))
      ) #end second column/main panel
    ), #end of FluidRow
    hr(),
    fluidRow(
      column(9, a(href="https://github.com/miraisolutions/swissretire.git",icon("github-square", "fa-2x"))),
      column(3, a(href="http://www.mirai-solutions.com", img(src='mirai.png', align = "right",  height = "20%", width = "40%")), align="right" )
      ,style = " margin-bottom: 10px;" 
      )
  ) # end of fluidPage
) #end of shinyUI